libgeos/libgeos_handle.go

```diff
@@ -0,0 +1,236 @@
+		return true, nil
+	}
+	return h.relate(g1, g2, "T*F**FFF*")
+}
+
+// Disjoint returns true if and only if the input geometries have no points in
+// common.
+func (h *Handle) Disjoint(g1, g2 geom.Geometry) (bool, error) {
+	return h.relate(g1, g2, "FF*FF****")
+}
+
+// Touches returns true if and only if the geometries have at least 1 point in
+// common, but their interiors don't intersect.
+func (h *Handle) Touches(g1, g2 geom.Geometry) (bool, error) {
+	return h.relatesAny(
+		g1, g2,
+		"FT*******",
+		"F**T*****",
+		"F***T****",
+	)
+}
+
+// Contains returns true if and only if geometry A contains geometry B.  See
+// the global Contains function for details.
+func (h *Handle) Contains(a, b geom.Geometry) (bool, error) {
+	return h.relate(a, b, "T*****FF*")
+}
+
+// Covers returns true if and only if geometry A covers geometry B. See the
+// global Covers function for details.
+func (h *Handle) Covers(a, b geom.Geometry) (bool, error) {
+	return h.relatesAny(
+		a, b,
+		"T*****FF*",
+		"*T****FF*",
+		"***T**FF*",
+		"****T*FF*",
+	)
+}
+
+// Intersects returns true if and only if the geometries share at least one
+// point in common.
+func (h *Handle) Intersects(a, b geom.Geometry) (bool, error) {
+	return h.relatesAny(
+		a, b,
+		"T********",
+		"*T*******",
+		"***T*****",
+		"****T****",
+	)
+}
+
+// Within returns true if and only if geometry A is completely within geometry
+// B. See the global Within function for details.
+func (h *Handle) Within(a, b geom.Geometry) (bool, error) {
+	return h.relate(a, b, "T*F**F***")
+}
+
+// CoveredBy returns true if and only if geometry A is covered by geometry B.
+// See the global CoveredBy function for details.
+func (h *Handle) CoveredBy(a, b geom.Geometry) (bool, error) {
+	return h.relatesAny(
+		a, b,
+		"T*F**F***",
+		"*TF**F***",
+		"**FT*F***",
+		"**F*TF***",
+	)
+}
+
+// relatesAny checks if the two geometries are related using any of the masks.
+func (h *Handle) relatesAny(g1, g2 geom.Geometry, masks ...string) (bool, error) {
+	for _, m := range masks {
+		r, err := h.relate(g1, g2, m)
+		if err != nil {
+			return false, err
+		}
+		if r {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+// relate invokes the libgeos GEOSRelatePattern function, which checks if two
+// geometries are related according to a DE-9IM 'relates' mask.
+func (h *Handle) relate(g1, g2 geom.Geometry, mask string) (bool, error) {
+	if g1.IsGeometryCollection() || g2.IsGeometryCollection() {
+		return false, ErrGeometryCollectionNotSupported
+	}
+
+	gh1, err := h.createGeometryHandle(g1)
+	if err != nil {
+		return false, err
+	}
+	defer C.GEOSGeom_destroy(gh1)
+
+	gh2, err := h.createGeometryHandle(g2)
+	if err != nil {
+		return false, err
+	}
+
+	defer C.GEOSGeom_destroy(gh2)
+	cmask := C.CString(mask)
+	defer C.free(unsafe.Pointer(cmask))
+
+	switch ret := C.GEOSRelatePattern_r(h.context, gh1, gh2, cmask); ret {
+	case 2:
+		return false, h.err()
+	case 1:
+		return true, nil
+	case 0:
```

Comment:
Should it use `const` name for the value instead of just value itself. Just checked the source code. What I could understand is that 2 means something goes wrong, 1 means ture and 0 means false. But that would be good to have a string name instead of value to help people to read it

Reply:
Good suggestion, thanks! Fixed.

Reply:
Good suggestion, thanks! Fixed.